src/components/SyncIndicator.tsx

```diff
@@ -184,181 +173,248 @@ export const SyncIndicator = ({ handle }: { handle: DocHandle<unknown> }) => {
+    }
+
+    if (arraysAreEqual(ownHeads, syncServerHeads)) {
+      send("IS_IN_SYNC");
+    } else {
+      send("IS_OUT_OF_SYNC");
+    }
+  }, [ownHeads, syncServerHeads]);
+
   return {
-    isSynced,
+    ownHeads,
+    syncServerHeads,
     lastSyncUpdate,
-    syncTimeoutTimestamp,
+    isInternetConnected: machine.matches("internet.connected"),
+    syncState: machine.matches("sync.unknown")
+      ? SyncState.Unknown
+      : machine.matches("sync.inSync")
+      ? SyncState.InSync
+      : SyncState.OutOfSync,
+
+    // todo: add reachability check, currently this value will be always true
+    syncServerConnectionError: machine.matches("syncServer.disconnected.error"),
+    syncServerResponseError: machine.matches("sync.outOfSync.error"),
   };
 }
 
-function useIsConnectedToServer() {
-  const repo = useRepo();
-  const [isConnected, setIsConnected] = useState(() =>
-    repo.peers.some((p) => p.match(SYNC_SERVER_PREFIX))
-  );
+interface SyncIndicatorMachineConfig {
+  // the duration we wait for the sync server to respond in the unsynced state before we show an error
+  // the timer starts once both internet.connected and sync.isOutOfSync become true
+  connectionInitTimeout: number;
 
-  useEffect(() => {
-    const onPeerConnected = ({ peerId }) => {
-      if (peerId.match(SYNC_SERVER_PREFIX)) {
-        setIsConnected(true);
-      }
-    };
+  // the duration we wait for the sync server to respond in the unsynced state before we show an error
+  // the timer starts once both internet.connected and sync.isOutOfSync become true
+  maxSyncMessageDelay: number;
 
-    const onPeerDisconnnected = ({ peerId }) => {
-      if (peerId.match(SYNC_SERVER_PREFIX)) {
-        setIsConnected(false);
-      }
-    };
-
-    repo.networkSubsystem.on("peer", onPeerConnected);
-    repo.networkSubsystem.on("peer-disconnected", onPeerDisconnnected);
-  }, [repo]);
+  // initial internet connection state
+  isInternetConnected?: boolean;
 
-  return isConnected;
-}
+  // initial sync server connection state
+  isSyncServerConnected?: boolean;
 
-interface OnlineState {
-  isOnline: boolean;
-  timestamp: number; // when the state (offline/online) was entered
+  // initial is sync state
+  isInSync?: boolean;
 }
 
-function useIsOnline(): OnlineState {
-  const [isOnlineState, setIsOnlineState] = useState<OnlineState>({
-    isOnline: navigator.onLine,
-    timestamp: Date.now(),
-  });
-
-  useEffect(() => {
-    const onOnline = () => {
-      setIsOnlineState({ isOnline: true, timestamp: Date.now() });
-    };
-
-    const onOffline = () => {
-      setIsOnlineState({ isOnline: false, timestamp: Date.now() });
-    };
-
-    window.addEventListener("online", onOnline);
-    window.addEventListener("offline", onOffline);
-
-    return () => {
-      window.removeEventListener("online", onOnline);
-      window.removeEventListener("offline", onOffline);
-    };
-  }, []);
-
-  return isOnlineState;
+export function getSyncIndicatorMachine({
```

Comment:
This is definitely an improvement! I wonder if we could go further with decoupling the logic from the React view. There are currently a lot of hooks and React state which need to interact with the state machine. What if we extract out all the sync logic to a stateful class which takes in the handle and keeps track of all the heads, and then the React component has more limited responsibilities?

(not sure whether this would be an improvement but curious what you think @paulsonnentag)

Reply:
I like that idea, will refactor this